8326381: com.sun.net.httpserver.HttpsParameters and SSLStreams incorrectly handle needClientAuth and wantClientAuth

jaikiran opened the pull request:
Can I please get a review of this change which proposes to fix https://bugs.openjdk.org/browse/JDK-8326381?

As noted in the JBS issue, the implementation in `setNeedClientAuth()` and `setWantClientAuth()` of `com.sun.net.httpserver.HttpsParameters` wasn't matching the API specification. The commit in this PR fixes that issue and it now matches the API specification as well as what is done in `javax.net.ssl.SSLParameters` class.

Additionally, as noted in the JBS issue, the (internal class) `sun.net.httpserver.SSLStreams` had a bug where it could end up resetting the `needClientAuth` flag on the `SSLEngine` because of the way the `setNeedClientAuth()` and `setWantClientAuth()` methods were being called on the `SSLEngine`. This too has been fixed in this PR.

A new jtreg test has been introduced to reproduce the issue in the `HttpsParameters` class and verify this fix.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8326413](https://bugs.openjdk.org/browse/JDK-8326413) to be approved

### Issues
 * [JDK-8326381](https://bugs.openjdk.org/browse/JDK-8326381): com.sun.net.httpserver.HttpsParameters and SSLStreams incorrectly handle needClientAuth and wantClientAuth (**Bug** - P4)
 * [JDK-8326413](https://bugs.openjdk.org/browse/JDK-8326413): com.sun.net.httpserver.HttpsParameters and SSLStreams incorrectly handle needClientAuth and wantClientAuth (**CSR**)


### Reviewers
 * [Daniel Jeliński](https://openjdk.org/census#djelinski) (@djelinski - **Reviewer**)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17940/head:pull/17940` \
`$ git checkout pull/17940`

Update a local copy of the PR: \
`$ git checkout pull/17940` \
`$ git pull https://git.openjdk.org/jdk.git pull/17940/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17940`

View PR using the GUI difftool: \
`$ git pr show -t 17940`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17940.diff">https://git.openjdk.org/jdk/pull/17940.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17940#issuecomment-1956010847)

jaikiran commented:
This change fixes the implementation of the `HttpsParameters` methods to match its API specification. At the same time, it's still a change in behaviour of these methods. Should we file a CSR for this change?

djelinski commented:
If I understand correctly, `com.sun.net.httpserver` is not part of the spec, and CSRs are for spec changes only.

Have you verified that the new code works as intended?
- sends a CertificateRequest message only if either of the properties is true
- refuses to connect if needClientAuth is true and the client doesn't produce a certificate

Pretty sure we don't have any tests for that, they couldn't possibly pass with the current code.

johnshajiang commented on src/jdk.httpserver/share/classes/sun/net/httpserver/SSLStreams.java:
```diff
@@ -85,8 +85,12 @@ private void configureEngine(HttpsConfigurator cfg, InetSocketAddress addr){
+                if (params.getNeedClientAuth()) {
+                    engine.setNeedClientAuth(true);
+                }
+                if (params.getWantClientAuth()) {
```
Assume the states of `wantClientAuth` and `needClientAuth` are correctly maintained with the changes in `HttpsParameters`, or it's impossible both of them are `true`.
Could here use `if-else if` clause, like the below?
```
if (params.getNeedClientAuth()) {
    engine.setNeedClientAuth(true);
} else if (params.getWantClientAuth()) {
    engine.setWantClientAuth(true);
}
```

johnshajiang commented on test/jdk/com/sun/net/httpserver/HttpsParametersClientAuthTest.java:
```diff
@@ -0,0 +1,94 @@
+     * as expected
+     */
+    @Test
+    public void testClientAuth() throws Exception {
```
Just a suggestion.

Now that this test uses JUnit, why doesn't it define multiple test methods for the different HttpsParameters instances?
A single test method just focus on only one HttpsParameters instance (or test case).

If the checking on a HttpsParameters instance (or test case) fails, the remaining test cases still can be executed.

jaikiran replied:
Hello John, I've updated the PR to split these into separate test methods.

johnshajiang commented on test/jdk/com/sun/net/httpserver/HttpsParametersClientAuthTest.java:
```diff
@@ -0,0 +1,94 @@
+                "needClientAuth was expected to be false but wasn't");
+    }
+
+
```
Nit: this blank line could be removed if you want.

johnshajiang commented on test/jdk/com/sun/net/httpserver/HttpsParametersClientAuthTest.java:
```diff
@@ -0,0 +1,94 @@
+                "wantClientAuth was expected to be true but wasn't");
+        assertFalse(wantClientAuthParams.getNeedClientAuth(),
+                "needClientAuth was expected to be false but wasn't");
+    }
```
Is it necessary to check the states on the following cases?

```
HttpsParameters params = new Params();
params.setNeedClientAuth(true);
params.setWantClientAuth(false);
```

```
HttpsParameters params = new Params();
params.setWantClientAuth(true);
params.setNeedClientAuth(false);
```

jaikiran replied:
I don't have preference, especially since we now decided to deprecate these methods. But if you prefer we add those, then let me know and I can update the test tomorrow.

johnshajiang replied:
I thought this fix made different states of `wantClientAuth` and `needClientAuth` for those cases.
So, it may be better to check if the states in those cases are expected, though they should be corner cases.

Michael-Mc-Mahon commented:
> If I understand correctly, `com.sun.net.httpserver` is not part of the spec, and CSRs are for spec changes only.
> 

This would be a JDK (scope) spec change, rather than a Java SE change as far as I remember.

jaikiran commented:
Given Michael's input, I'll file a CSR for this change. I'll address the review comments too shortly.

/csr needed

Michael-Mc-Mahon commented:
> If I understand correctly, `com.sun.net.httpserver` is not part of the spec, and CSRs are for spec changes only.
> 
> Have you verified that the new code works as intended?
> 
>     * sends a CertificateRequest message only if either of the properties is true
> 
>     * refuses to connect if needClientAuth is true and the client doesn't produce a certificate
> 
> 
> Pretty sure we don't have any tests for that, they couldn't possibly pass with the current code.

On. the second point there, I think it would be useful if we had a test for this. It could be done in another PR maybe, but it would need a client/server interaction with the "need" flag set and if no client cert available, check for appropriate error. If cert available the client and server can both check that it was used, through the SSLSession created.

You could use the same approach to test the "want" flag as well potentially.

Michael-Mc-Mahon commented:
> Should we deprecate the HttpsParameters methods? The class javadoc has advised against using them since the day they were introduced, and if I read [JDK-6398764](https://bugs.openjdk.org/browse/JDK-6398764) correctly, they were only introduced to enable backporting HttpsParameters to JDK 1.5.

Yes, I'd agree with this.

jaikiran commented:
I've updated this PR to deprecate the methods on `HttpsParameters` which were meant to set some SSL parameters. 

Additionally, I have also added more tests, this time to verify that if a server is configured using `HttpsParameters.setNeedClientAuth()` and `HttpsParameters.setWantClientAuth()` then the TLS handshake works as expected and fails in the case when the client doesn't present a certificate when needClientAuth is true. I have verified that without the fix source fix, the test fails and after the fix the test passes.

I will focus on the CSR text including the deprecation parts tomorrow once we have an agreement on these changes.

jaikiran commented:
Hello Michael,

> On. the second point there, I think it would be useful if we had a test for this. It could be done in another PR maybe, but it would need a client/server interaction with the "need" flag set and if no client cert available, check for appropriate error. If cert available the client and server can both check that it was used, through the SSLSession created.

> You could use the same approach to test the "want" flag as well potentially.

I did add a test to verify this semantic. However, I took a slightly different approach in the test. When the server is configured with needClientAuth, I created a client with a `SSLContext` which doesn't have any key material (and thus cannot present a valid certificate) and only has trust material (and thus can trust the server certificate). That client then initiated a HTTP request and the test verifies that such a request fail during TLS handshake (since the client doesn't present the certificate).

The wantClientAuth test too does the same but in that case the request is expected to pass even when the client can't present the valid certificate.

Given the context of this issue, do you think this way of testing is reasonable?

johnshajiang commented on test/jdk/com/sun/net/httpserver/HttpsParametersClientAuthTest.java:
```diff
@@ -0,0 +1,379 @@
+        // first set wantClientAuth to an initial value to verify that it later gets reset
+        needClientAuthParams.setWantClientAuth(initialWantClientAuth);
+        // needClientAuth = true and thus wantClientAuth = false
+        needClientAuthParams.setNeedClientAuth(true);
```
I just want to clarify that these new cases look not suggested by my previous comment.

In my suggested cases, as shown as the below, 
```
setNeedClientAuth(true);
setWantClientAuth(false);
```
```
setWantClientAuth(true);
setNeedClientAuth(false);
```
this first setting takes a state (`needClientAuth` or `wantClientAuth`) to `true`, and the second setting takes both states to `false`.
Before this fix, the second setting doesn't reset the `true` state specified by the first setting to `false`.

Of course, if you think it's unnecessary to cover these corner cases, please feel free to ignore them.

jaikiran replied:
Hello John, it wasn't clear to me previously that this is the specific test you were suggesting. I have now updated the PR to introduce this additional test.

vbrandl reviewed (commented):
The fix for [8326233](https://bugs.openjdk.org/browse/JDK-8326233) uses two consecutive `if`s instead of `if`/`else if`: https://github.com/openjdk/jdk/pull/17923/files#diff-6fdccf370b8c2ed3fc9690a3f5ee9a7d65af933a84252d99709c8926c52de43dR628

This shouldn't make a difference except for invalid parameter objects (which could be created using reflection or by extending the class). I'm just asking if both places should use the same logic when copying the parameters?

Also (and this is out of scope for this PR) maybe it would be better to deprecate the separate want/need client auth fields in favor of an enum `enum ClientAuth { NONE, WANT, NEED }` and only one setter `setClientAuth`. This should be less error prone

jaikiran commented:
Hello Valentin,

> The fix for [8326233](https://bugs.openjdk.org/browse/JDK-8326233) uses two consecutive `if`s instead of `if`/`else if`: https://github.com/openjdk/jdk/pull/17923/files#diff-6fdccf370b8c2ed3fc9690a3f5ee9a7d65af933a84252d99709c8926c52de43dR628
> 
> This shouldn't make a difference except for invalid parameter objects (which could be created using reflection or by extending the class). I'm just asking if both places should use the same logic when copying the parameters?

Like you note it shouldn't make a difference functionally, especially with the tests we have now added. Ideally, using if/else in that other part of the code would have more clearly indicated the intention and expectation of that code. When we next touch that `HttpClient` code, I will update that part to use `if/else` to be consistent.

> Also (and this is out of scope for this PR) maybe it would be better to deprecate the separate want/need client auth fields in favor of an enum enum ClientAuth { NONE, WANT, NEED } and only one setter setClientAuth. This should be less error prone 

A similar suggestion has been made in https://bugs.openjdk.org/browse/JDK-8311114. It would need inputs from security-libs team about that proposed API.

jaikiran commented:
The implementation changes, the deprecation and the test coverage changes are now complete. I will focus on the CSR text next.

jaikiran commented on test/jdk/com/sun/net/httpserver/HttpsParametersClientAuthTest.java:
```diff
@@ -0,0 +1,513 @@
+                        // verify it failed due to right reason
+                        Throwable cause = ioe;
+                        while (cause != null) {
+                            // either of SocketException or SSLHandshakeException are OK
```
I have this on my TODO list and I plan to come back to this shortly after I am done with the CSR text. I was expecting this to be one specific type of exception but I'm seeing 2 different types depending on how I launch the test. Plus there's an open PR in security-libs area which is proposing some changes in the alert/exception raised from the TLS handshake layer when the client doesn't send certificates to the server. I need to review those changes to see what kind of impact if any that change will have here.

djelinski commented on test/jdk/com/sun/net/httpserver/HttpsParametersClientAuthTest.java:
```diff
@@ -0,0 +1,513 @@
+
+        @Override
+        public X509Certificate[] getCertificateChain(String alias) {
+            clientAuthInitiated = true; // keep track that client certs was requested
```
nit: this is not necessary - `getCertificateChain` is only called if `choose*Alias` returns a non-null value, and both the `chooseClientAlias` methods set `clientAuthInitiated = true` already.

djelinski reviewed (approved):
LGTM.

jaikiran commented:
/integrate
